[Improvement] Sanitize YAML-derived values before embedding in curl JSON payloads

## Objective
Ensure all YAML-derived string values are properly JSON-escaped before being passed to \`curl -d\` in API calls, preventing JSON injection and command injection attacks.

## Motivation
Issue #20 identified that \`apply.sh\` passes unsanitized YAML values to curl JSON body. While \`build_create_json\` in \`reconcile.sh\` uses \`jq -n --arg\` (which safely escapes), the \`patch_body\` construction in \`apply.sh\` line 203-210 also uses \`jq -n --arg\` — however, any future code that constructs JSON strings manually (e.g., via string interpolation) would be vulnerable. This issue tracks adding a defense-in-depth sanitization layer and audit of all JSON construction paths.

## Deliverables
- [ ] Audit all JSON construction in \`apply.sh\`, \`reconcile.sh\`, \`api.sh\`, and \`export.sh\`
- [ ] Create \`scripts/lib/sanitize.sh\` with a \`json_escape\` function
- [ ] Add validation: reject YAML values containing null bytes or control characters
- [ ] Add \`shellcheck\` SC2086 compliance for all variable expansions in curl commands
- [ ] Add integration test: agent with special characters in taskDescription (quotes, backslashes, newlines)
- [ ] Ensure \`export.sh\` YAML generation properly quotes all string values

## Success Criteria
- [ ] An agent with \`taskDescription: "He said \"hello\" and \n left"\` roundtrips correctly through export → apply
- [ ] No JSON construction uses string interpolation without jq escaping
- [ ] shellcheck passes with no SC2086 warnings in API-related code

## Priority
HIGH -- Injection attacks via YAML values could compromise the ai-maestro control plane

# Implementation Plan

Now I have a thorough picture of the codebase. Let me produce the implementation plan.

---

# Implementation Plan: Issue #53 — Sanitize YAML-derived Values in curl JSON Payloads

## 1. Objective

Add defense-in-depth sanitization for all YAML-derived string values before they reach `curl` JSON payloads. The primary risks are:

- **YAML injection in `export.sh`**: unquoted heredoc fields that contain YAML-special characters (`:`  `#`  `{`  etc.) produce malformed YAML on the next export/import cycle.
- **Null byte / control character injection**: a crafted YAML value containing `\x00`–`\x1f` could corrupt JSON payloads or bypass validation.
- **No ShellCheck enforcement**: SC2086 unquoted expansions could creep in through future changes.
- **No integration test coverage**: special-character roundtrip is untested.

Existing `jq -n --arg` usage in `apply.sh` and `reconcile.sh` is already safe for JSON construction. The plan adds a validation layer and fixes the weaker YAML-quoting in `export.sh`.

---

## 2. Approach

| Decision | Rationale |
|---|---|
| Central `sanitize.sh` library | Single place to add/update rules; sourced by all scripts |
| Validate at read time, not write time | Fail fast when values are first loaded from YAML, before any downstream JSON/YAML construction |
| `jq -Rr` → `jq --rawfile` for YAML string quoting | `jq -Rr '.'` strips JSON encoding but does NOT YAML-quote; replace with a proper YAML string quoting function |
| No regex for escaping — use `jq` and `printf %q` | Canonical tools beat regex (lesson from team KB) |
| ShellCheck via `.shellcheckrc` at repo root | Enforce SC2086 without suppressing it; scope per-directory overrides in sub-`.shellcheckrc` files if needed |
| BATS for integration tests | Standard shell testing framework; mock `curl` / `jq` via PATH prepend |

---

## 3. Files to Create

### `scripts/lib/sanitize.sh`
Purpose: Provides `json_escape`, `yaml_quote_string`, and `validate_safe_value` functions. Sourced by all other scripts.

Key functions:
- `validate_safe_value(name, value)` — rejects null bytes and raw control characters (`\x00`–`\x1f` excluding `\t` `\n` `\r`); prints a diagnostic and returns 1 on violation.
- `json_escape(value)` — wraps `jq -Rn --arg v "$value" '$v'`; emits a properly JSON-encoded string. Use as a belt-and-suspenders check before `--arg` calls.
- `yaml_quote_string(value)` — emits the value wrapped in YAML double-quotes with internal `"` and `\` escaped (using `printf` and `sed`), suitable for safe heredoc embedding.

### `tests/shell/test_sanitize.bats`
Purpose: BATS unit tests for `sanitize.sh` functions.

Covers:
- `validate_safe_value` rejects null bytes, form-feed, vertical-tab
- `validate_safe_value` accepts tab, newline, carriage-return, printable ASCII, multi-byte UTF-8
- `json_escape` round-trips quotes, backslashes, newlines
- `yaml_quote_string` produces YAML-parseable output for quotes/backslashes/newlines/colons

### `tests/shell/test_roundtrip.bats`
Purpose: Integration test — agent with special characters in `taskDescription` roundtrips through export → apply without data loss or injection.

Covers the success criterion: `taskDescription: 'He said "hello" and \n left'` roundtrips correctly.

Uses mock `curl` stub that captures the `-d` body for assertion.

### `tests/shell/helpers/common.bash`
Purpose: Shared BATS setup helpers — `setup_mocks()`, `clean_state()`, PATH injection for mock binaries.

### `tests/shell/mocks/curl`
Purpose: Mock `curl` binary. Writes received `-d` argument to `$MOCK_CURL_CAPTURE_FILE` for inspection in tests.

### `.shellcheckrc` (repo root)
Purpose: Enable `SC2086` (unquoted expansions) and `SC2181` (check exit codes directly) globally.

---

## 4. Files to Modify

### `scripts/lib/reconcile.sh`

| Location | Change |
|---|---|
| Top of file, after `set -euo pipefail` | `source "$(dirname "${BASH_SOURCE[0]}")/sanitize.sh"` |
| `build_create_json()` — before jq call | Call `validate_safe_value` for each of the nine arguments |
| Tag CSV split (line ~80) | Add `validate_safe_value "tags" "$tags_csv"` before the `jq -Rc 'split(",")'` call |

### `scripts/apply.sh`

| Location | Change |
|---|---|
| Top of file, after existing `source` lines | `source "${SCRIPT_DIR}/lib/sanitize.sh"` |
| Before patch body construction (~line 200) | Call `validate_safe_value` for `$label`, `$program`, `$workdir`, `$args`, `$desc` — these are already read from YAML; validate here before any API call |
| Quoting audit | Verify all `curl` arguments (`"$actual_id"`, `"$patch_body"`) are quoted — fix any SC2086 hits |

### `scripts/export.sh`

| Location | Change |
|---|---|
| Top of file | `source "$(dirname "${BASH_SOURCE[0]}")/lib/sanitize.sh"` |
| Lines 131–141 (per-field `jq -Rr`) | Replace bare `jq -Rr '.'` pipe with `yaml_quote_string` for every field that is embedded unquoted in the heredoc (`name_yaml`, `label_yaml`, `program_yaml`, `workdir_yaml`, `owner_yaml`, `role_yaml`) |
| Heredoc (lines 143–166) | Wrap the model/state/docker fields that are currently unquoted with `yaml_quote_string` — or keep literal-safe fields as-is and document why |
| `$args_escaped` / `$desc_escaped` | These are already in heredoc double-quotes; replace `jq -Rr` with `yaml_quote_string` to properly escape embedded `"` and `\` inside the YAML double-quoted string |

### `scripts/lib/api.sh`

| Location | Change |
|---|---|
| `agamemnon_create_agent()`, `agamemnon_update_agent()`, `agamemnon_docker_create()` | Confirm `"${body}"` is quoted (SC2086 audit) — it is in the current code, but shellcheck will confirm |
| Header comment | Add note: "JSON bodies must be pre-built via `jq --arg`/`--argjson`; never interpolate shell variables directly" |

### `scripts/plan.sh` and `scripts/status.sh`

| Location | Change |
|---|---|
| Any YAML value reads passed to output | Audit for unquoted expansions; these scripts are read-only but SC2086 still applies |

### `.github/workflows/validate.yml`

| Location | Change |
|---|---|
| Add step | Run `shellcheck` on `scripts/**/*.sh` — install via `apt-get install shellcheck` or use the `shellcheck-action` |

---

## 5. Implementation Order

1. **Create `scripts/lib/sanitize.sh`** with `validate_safe_value`, `json_escape`, and `yaml_quote_string`. This is the foundation; nothing else can be tested without it.

2. **Create test infrastructure** — `tests/shell/helpers/common.bash`, `tests/shell/mocks/curl` (chmod +x). BATS must be runnable before writing tests.

3. **Write `tests/shell/test_sanitize.bats`** — unit tests for all three `sanitize.sh` functions. Green-light the library before wiring it in.

4. **Modify `scripts/export.sh`** — source `sanitize.sh`; replace `jq -Rr` with `yaml_quote_string`; fix all unquoted heredoc variables. This fixes the highest-risk path (YAML round-trip corruption).

5. **Modify `scripts/lib/reconcile.sh`** — source `sanitize.sh`; add `validate_safe_value` calls in `build_create_json` and before tag CSV split.

6. **Modify `scripts/apply.sh`** — source `sanitize.sh`; add `validate_safe_value` calls before patch body construction; fix any SC2086 issues.

7. **Audit `scripts/lib/api.sh`** — confirm quoting; add documentation comment; fix any SC2086 hits.

8. **Write `tests/shell/test_roundtrip.bats`** — the integration test with special-character `taskDescription`. Uses the mock `curl` stub. Must pass after steps 4–7.

9. **Create `.shellcheckrc`** at repo root enabling SC2086, SC2181.

10. **Update `.github/workflows/validate.yml`** — add `shellcheck` step.

11. **Run full test suite** and shellcheck; iterate until clean.

---

## 6. Verification

### Unit tests (BATS)
```bash
# Install BATS if needed
npm install -g bats   # or: apt-get install bats

# Run unit tests
bats tests/shell/test_sanitize.bats

# Run integration tests
bats tests/shell/test_roundtrip.bats
```

### ShellCheck
```bash
shellcheck scripts/*.sh scripts/lib/*.sh
# Should produce zero SC2086 warnings
```

### Manual roundtrip (success criterion from issue)
```bash
# Create test agent YAML with special taskDescription
cat > /tmp/test-agent.yaml <<'EOF'
apiVersion: myrmidons/v1
kind: Agent
metadata:
  name: test-special
  host: hermes
spec:
  label: "Test Agent"
  program: claude-code
  model: null
  workingDirectory: /tmp
  programArgs: ""
  taskDescription: "He said \"hello\" and \n left"
  tags: [test]
  owner: mvillmow
  role: member
  deployment:
    type: local
  desiredState: active
EOF

# Apply (with mock Agamemnon or real), then export, then diff
./scripts/apply.sh
./scripts/export.sh
yq eval '.spec.taskDescription' agents/hermes/test-special.yaml
# Must output: He said "hello" and \n left
```

### Null byte rejection test
```bash
# Confirm validate_safe_value rejects control characters
source scripts/lib/sanitize.sh
validate_safe_value "desc" $'malicious\x00value'
# Must return exit code 1 and print an error
```

### CI gate
After PR merge, `.github/workflows/validate.yml` runs shellcheck + BATS; PR cannot merge if either fails.

---

## 7. Skills Used

### Skills Invoked During Planning
- **`hephaestus:advise`** — searched team knowledge base for prior learnings on JSON injection, shellcheck, BATS, and YAML quoting patterns before drafting the plan.
- **`Explore` subagent** — performed codebase analysis of `apply.sh`, `reconcile.sh`, `api.sh`, `export.sh`, and `tests/` to understand current JSON construction patterns and security posture.

### Team Knowledge Base Skills Referenced (from Prior Learnings)
| Skill | How It Influenced This Plan |
|---|---|
| `jq-compatibility-fix` | Confirmed `jq -n --arg`/`--argjson` as canonical safe pattern; `build_create_json` already follows it; plan preserves and enforces it |
| `shellcheck-scope-templates` | Drives the `.shellcheckrc` at repo root + per-subdirectory override approach |
| `bats-shell-testing` | Drives the entire test structure: `tests/shell/`, `mocks/` with PATH prepend, `helpers/common.bash`, env-var-controlled mock stubs |
| `github-actions-ci-patterns` | Drives the `validate.yml` shellcheck step; the `env:` block analogy reinforces always routing values through `jq --arg` |
| `fix-yaml-frontmatter-colon-bug` | Directly informed the `export.sh` fix — colon-in-value truncation is why `jq -Rr '.'` alone is insufficient for YAML quoting |
| `security-review-false-positive-filter` | Guided the audit methodology: trace full call chains before declaring a pattern safe or unsafe |
| `preflight-script-integration-patterns` | Reinforces `set -uo pipefail` header and `|| true` on grep calls in new scripts |

---
*Generated by Claude Code Planner*

--- a/scripts/apply.sh
+++ b/scripts/apply.sh
@@ -60,6 +60,7 @@ HIBERNATED=0
 UNCHANGED=0
 PRUNED=0
 ERRORS=0
+FAILED_AGENTS=()   # names of agents that encountered errors
 
 # Per-agent error tracking: parallel arrays of (agent_name, http_status, error_message)
 FAILED_AGENT_NAMES=()
@@ -350,7 +351,15 @@ main() {
     else
         echo ""
         echo "================================================"
-        echo "Summary: created=${CREATED} updated=${UPDATED} woken=${WOKEN} hibernated=${HIBERNATED} unchanged=${UNCHANGED} errors=${ERRORS}"
+        echo "Summary: created=${CREATED} updated=${UPDATED} woken=${WOKEN} hibernated=${HIBERNATED} unchanged=${UNCHANGED} pruned=${PRUNED} errors=${ERRORS}"
+
+        if [[ ${#FAILED_AGENTS[@]} -gt 0 ]]; then
+            echo ""
+            echo "Failed agents (${#FAILED_AGENTS[@]}):"
+            for fa in "${FAILED_AGENTS[@]}"; do
+                echo "  - ${fa}"
+            done
+        fi
 
         # Always save a report file (silently) so report_save is useful even in text mode
         local report_json
@@ -453,6 +462,7 @@ apply_agent() {
                 echo "    ERROR creating ${name}: ${err_msg}" >&2
             fi
             record_failure "$name" "$http_status" "$err_msg"
+            FAILED_AGENTS+=("$name")
             report_add_agent "$name" "$agent_host" "ERROR" "$desired_state" "unknown" "[]" "create failed: ${err_msg}"
         fi
         return
@@ -533,6 +543,8 @@ apply_agent() {
             fi
 
             local patch_body
+            # Note: $label is a reserved keyword in jq 1.6 (label-break syntax).
+            # Use $lbl as the variable name to avoid the parser conflict.
             patch_body="$(jq -n \
                 --arg lbl "$label" \
                 --arg program "$program" \
@@ -572,6 +584,7 @@ apply_agent() {
                     echo "    ERROR updating ${name}: ${err_msg}" >&2
                 fi
                 record_failure "$name" "$http_status" "Failed to update fields [${changed_fields}]: ${err_msg}"
+                FAILED_AGENTS+=("$name")
                 report_add_agent "$name" "$agent_host" "ERROR" "$desired_state" "$actual_status" "$drift_json" "update failed: ${err_msg}"
             fi
             ;;

--- a/scripts/export.sh
+++ b/scripts/export.sh
@@ -60,8 +60,17 @@ main() {
 }
 
 check_jq() {
+    local missing=()
     if ! command -v jq &>/dev/null; then
-        log_error "jq is required. Install: apt install jq"
+        missing+=("jq")
+    fi
+    if ! command -v yq &>/dev/null; then
+        missing+=("yq")
+    fi
+    if [[ ${#missing[@]} -gt 0 ]]; then
+        log_error "Missing required tools: ${missing[*]}"
+        log_error "  Install jq: apt install jq / brew install jq"
+        log_error "  Install yq: https://github.com/mikefarah/yq"
         exit 1
     fi
 }

